### PR TITLE
Circuit range constraints

### LINKED ISSUE #417: Missing Range Checks on Comparison Circuits
Some computations, such as comparisons, cannot be directly expressed as constraints in Circom. Instead, CircomLib provides templates that developers can use so that they don’t have to implement them from scratch. Importantly, some of these templates make assumptions about the inputs that the developers must enforce. The `LessThan(N)`, `LessEqThan(N)`, `GreaterThan(N)` and `GreaterEqThan(N)` templates are all examples of such library templates as they all assume that the inputs are all constrained to be within `N` bits. If this does not hold, the template output can be manipulated. As an example, consider the snippet below.

> Snippet of `EpochKeyLite` with an unsafe use of `LessThan`
> 
> 
> ```
> template EpochKeyLite(EPOCH_KEY_NONCE_PER_EPOCH) {
>     ...
> 
>     component nonce_lt = LessThan(8);
>     nonce_lt.in[0] <== nonce;
>     nonce_lt.in[1] <== EPOCH_KEY_NONCE_PER_EPOCH;
>     nonce_lt.out === 1;
> 
>     ...
> }
> ```
> 

The above code does not include a check that `nonce` is within the range of an 8 bit value. This allows a malicious user to manipulate the circuit by providing a very large `nonce`, say `P - 1`, causing an overflow in the `LessThan` circuit (where `P` is the large prime). This will result in the template improperly reporting that  `P - 1` is less than `EPOCH_KEY_NONCE_PER_EPOCH`.

## Impact

Several templates make extensive use of the comparator templates from CircomLib, including `ProveReputation`, `UserStateTransition` and `EpochKeyLite`. These circuits implement critical behaviors in the project and can be manipulated in ways the developers likely do not intend. For example, the above snippet is used to ensure that a user can only generate a limited number of keys within an epoch. Without the range check, users can generate about 255 more keys than intended.

## Recommendation

While a previous auditor recommended that some of these range checks should be removed to save constraints, their removal compromises the security of the system, especially when the inputs are private (as is the case with `nonce` above). Ensure that a range check is performed on all inputs to the comparator circuits.

### LINKED ISSUE #418: Potential overflow when proving reputation
The `ProveReputation` template allows users to prove that their reputation is sufficient to meet certain conditions set by protocols. For example, this circuit allows users to prove that their reputation exceeds some lower-bound or that it is less than some upper-bound. As these checks are performed, however, the circuit has the potential to overflow since there are no checks on the range of `data[0]` or `data[1]`. 

> Location where arithmetic may overflow in `ProveReputation`
> 
> 
> ```solidity
> template ProveReputation(STATE_TREE_DEPTH, EPOCH_KEY_NONCE_PER_EPOCH, SUM_FIELD_COUNT, FIELD_COUNT) {
>     ...
> 
>     component min_rep_check = GreaterEqThan(252);
>     min_rep_check.in[0] <== data[0];
>     min_rep_check.in[1] <== data[1] + min_rep;
> 
>     ...
> 
>     component max_rep_check = GreaterEqThan(252);
>     max_rep_check.in[0] <== data[1];
>     max_rep_check.in[1] <== data[0] + max_rep;
> 
>     ...
> }
> ```
> 

## Impact

If the “zero reputation” is chosen poorly or if the user has high enough reputation, it is possible for an overflow to allow users to prove that they meet a condition when they do not. Additionally, since it appears that both `data[0]` and `data[1]` are private, such errors could not be detected by the protocol.

## Recommendation

Perform appropriate range checks on `data[0]` and `data[1]` to ensure that an overflow does not occur.

### LINKED ISSUE #419: Unnecessary Num2Bits(254) constraints
Unlike most programming languages, operations in ZK circuits are made over signals that range from `0` to `P` where `P` is the large prime. For this reason, it is important to perform range checks using `Num2Bits` to ensure a signal only accepts values of a particular bit width. The developers have such range check, but rather than checking the bit width using `Num2Bits(N)` where `N` is the number of bits, they instead always use `Num2Bits(254)` and then check that the upper `254 - N` bits are 0 as shown below.

> An instance where `Num2Bits(254)` can be be replaced with `Num2Bits(252)`
> 
> 
> ```solidity
> template Modulo() {
>     ...
> 
>     // check that remainder and divisor are both < 2**252
>     component remainder_bits = Num2Bits(254);
>     remainder_bits.in <== remainder;
>     for (var x = 252; x < 254; x++) {
>         remainder_bits.out[x] === 0;
>     }
> 
>     ...
> }
> ```
> 

## Impact

As this pattern is used to perform range checks on specific values, it saves constraints to perform `Num2Bits` with the desired number of bits.

## Recommendation

Replace cases that use the above pattern with `Num2Bits` of the appropriate bit-width. 

For convenience, the required changes are provided below.

```solidity
1. epochKeyLite.circom

L33~L37
```
    component attester_id_bits = Num2Bits(254);
    attester_id_bits.in <== attester_id;
    for (var x = 160; x < 254; x++) {
        attester_id_bits.out[x] === 0;
    }
```
-->
```
    component attester_id_bits = Num2Bits(160);
    attester_id_bits.in <== attester_id;
```

L39~L43
```
    component epoch_bits = Num2Bits(254);
    epoch_bits.in <== epoch;
    for (var x = 64; x < 254; x++) {
        epoch_bits.out[x] === 0;
    }
```
-->
```
    component epoch_bits = Num2Bits(64);
    epoch_bits.in <== epoch;
```

2. modulo.circom

L20~L24
```
    component remainder_bits = Num2Bits(254);
    remainder_bits.in <== remainder;
    for (var x = 252; x < 254; x++) {
        remainder_bits.out[x] === 0;
    }
```
-->
```
    component remainder_bits = Num2Bits(252);
    remainder_bits.in <== remainder;
```

L26~L30
```
    component divisor_bits = Num2Bits(254);
    divisor_bits.in <== divisor;
    for (var x = 252; x < 254; x++) {
        divisor_bits.out[x] === 0;
    }
```
-->
```
    component divisor_bits = Num2Bits(252);
    divisor_bits.in <== divisor;
```

3. proveReputation.circom

L68~L72
```
    component min_rep_bits = Num2Bits(254);
    min_rep_bits.in <== min_rep;
    for (var x = 64; x < 254; x++) {
        min_rep_bits.out[x] === 0;
    }
```
-->
```
    component min_rep_bits = Num2Bits(64);
    min_rep_bits.in <== min_rep;
```

L74~L78
```
    component max_rep_bits = Num2Bits(254);
    max_rep_bits.in <== max_rep;
    for (var x = 64; x < 254; x++) {
        max_rep_bits.out[x] === 0;
    }
```
-->
```
    component max_rep_bits = Num2Bits(64);
    max_rep_bits.in <== max_rep;
```
```

### PR DESCRIPTION
**Pull request by vimwitch**
## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been reviewed and added / updated if needed (for bug fixes / features)
- [x] Build (`yarn build`) was run locally and any changes were pushed
- [x] Lint (`yarn lint --check`) has passed locally and any fixes were made for failures

## Pull request type

Please check the type of change your PR introduces:

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## What is the current behavior?

Closes #419 
Closes #418 
Closes #417 

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: packages/circuits/circuits/proveReputation.circom [hunk 1]
```diff
@@ -128,7 +134,7 @@ template ProveReputation(STATE_TREE_DEPTH, EPOCH_KEY_NONCE_PER_EPOCH, SUM_FIELD_
     /* 3. Check if user has reputation less than max_rep */
     // if proving max_rep > 0, check if data[1] >= data[0] + max_rep
 
-    component max_rep_check = GreaterEqThan(252);
+    component max_rep_check = GreaterEqThan(66);
```
Comment (vivianjeng) [inline packages/circuits/circuits/proveReputation.circom:137]: why a `66` here?

**Review by vivianjeng — APPROVED** [review]
LGTM
we should also check 
```js
assert(SUM_FIELD_COUNT < FIELD_COUNT);
```
or maybe
```js
assert(STATE_TREE_DEPTH <= 32);
assert(EPOCH_TREE_DEPTH <= 32);
assert(HISTORY_TREE_DEPTH <= 32);
```

File: packages/circuits/circuits/proveReputation.circom (hunk 1 above)
Comment (vimwitch) [reply packages/circuits/circuits/proveReputation.circom:137]: The input is `data[0] + max_rep`. Both of these are 64 bits, so `2^64 + 2^64 = 2^65`. I added one more bit for safety to get 66.

**Comment by vimwitch** [comment]
Yes, this PR doesn't close #420

**Pull request merged**